### PR TITLE
New React Native template that demonstrates deferred login

### PR DESCRIPTION
**Pull request by wmathurin**
When app first load: you see a login button in the navigation bar.
After logging in, that button becomes a logout button (and a list of contacts is displayed below).
After logging out, one is back to the initial state.

NB: logout will only work with changes coming to the SalesforceMobileSDK-ReactNative repo.

### CODE REVIEW

File: ReactNativeDeferredTemplate/app.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    View,
+    FlatList,
+    Button
+} from 'react-native';
+
+import { NavigationContainer } from '@react-navigation/native';
+import { createStackNavigator } from '@react-navigation/stack';
+import {oauth, net} from 'react-native-force';
+
+class ContactListScreen extends React.Component {
```
Comment (wmathurin) [inline ReactNativeDeferredTemplate/app.js:40]: This is the only file of interest.

**Comment by wmathurin** [comment]
The new app will only work once https://github.com/forcedotcom/SalesforceMobileSDK-ReactNative/pull/191 is merged.

File: ReactNativeDeferredTemplate/.flowconfig [hunk 2]
```diff
@@ -0,0 +1,58 @@
+[ignore]
+
+# We fork some components by platform.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/.flowconfig:3]: Are these files meant to be checked in or a part of `.gitignore`?

File: ReactNativeDeferredTemplate/android/app/src/main/java/com/salesforce/reactnativedeferredtemplate/MainApplication.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2015-present, salesforce.com, inc.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/android/app/src/main/java/com/salesforce/reactnativedeferredtemplate/MainApplication.java:2]: `2020-present`.

File: ReactNativeDeferredTemplate/android/keystores/debug.keystore.properties [hunk 4]
```diff
@@ -0,0 +1,4 @@
+key.store=debug.keystore
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/android/keystores/debug.keystore.properties:1]: I don't think this should be checked in, right?

File: ReactNativeDeferredTemplate/app.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2017-present, salesforce.com, inc.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/app.js:2]: `2020-present`.

File: ReactNativeDeferredTemplate/index.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2017-present, salesforce.com, inc.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/index.js:2]: `2020-present`.

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+			productType = "com.apple.product-type.application";
+		};
+/* End PBXNativeTarget section */
+
+/* Begin PBXProject section */
+		83CBB9F71A601CBA00E9B192 /* Project object */ = {
+			isa = PBXProject;
+			attributes = {
+				LastUpgradeCheck = 1020;
+				ORGANIZATIONNAME = Facebook;
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate.xcodeproj/project.pbxproj:150]: Should we change this to `salesforce.com`?

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/*
+ Copyright (c) 2015-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.h:2]: `2020-present`.

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m [hunk 9]
```diff
@@ -0,0 +1,175 @@
+/*
+ Copyright (c) 2011-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m:2]: `2020-present`.

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+{
+    self = [super init];
+    if (self) {
+      // Need to use SalesforceReactSDKManager in Salesforce Mobile SDK apps using React Native
+      [SalesforceReactSDKManager initializeSDK];
+      
+      //App Setup for any changes to the current authenticated user
+      [SFSDKAuthHelper registerBlockForCurrentUserChangeNotifications:^{
+        [self resetViewState:^{
+            [self setupRootViewController];
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m:52]: `strong-weak`?

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    // If you wish to register for push notifications, uncomment the line below.  Note that,
+    // if you want to receive push notifications from Salesforce, you will also need to
+    // implement the application:didRegisterForRemoteNotificationsWithDeviceToken: method (below).
+//    [self registerForRemotePushNotifications];
+    
+    //Uncomment the code below to see how you can customize the color, textcolor, font and fontsize of the navigation bar
+//    [self customizeLoginView];
+    
+    [SFSDKAuthHelper loginIfRequired:^{
+        [self setupRootViewController];
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m:73]: `strong-weak`?

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+- (BOOL)enableIDPLoginFlowForURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
+     return [[SFUserAccountManager sharedInstance] handleIDPAuthenticationResponse:url options:options];
+}
+#pragma mark - Private methods
+
+- (void)initializeAppViewState
+{
+    if (![NSThread isMainThread]) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self initializeAppViewState];
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m:139]: `strong-weak`?

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/InitialViewController.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
+/*
+ Copyright (c) 2013-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/InitialViewController.h:2]: `2020-present`.

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/InitialViewController.m [hunk 14]
```diff
@@ -0,0 +1,50 @@
+/*
+ Copyright (c) 2013-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/InitialViewController.m:2]: `2020-present`.

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/main.m [hunk 15]
```diff
@@ -0,0 +1,33 @@
+/*
+ Copyright (c) 2015-present, salesforce.com, inc. All rights reserved.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/main.m:2]: `2020-present`.

File: ReactNativeDeferredTemplate/template.js [hunk 16]
```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2016-present, salesforce.com, inc.
```
Comment (bhariharan) [inline ReactNativeDeferredTemplate/template.js:2]: `2020-present`.

**Comment by wmathurin** [comment]
@bhariharan Except for app.js changes, the switch from shouldAuthenticate true to false, it's a copy of of ReactNativeTemplate.

File: ReactNativeDeferredTemplate/.flowconfig (hunk 2 above)
Comment (wmathurin) [reply ReactNativeDeferredTemplate/.flowconfig:3]: There are meant to be checked in - they are part of ReactNativeTemplate also. They appear in the project generated with react native cli (or at least they did they last time I ran the react native cli).

File: ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m (hunk 12 above)
Comment (wmathurin) [reply ReactNativeDeferredTemplate/ios/ReactNativeDeferredTemplate/AppDelegate.m:139]: It's in all our templates ;-)
I don't think we need to worry about a retain cycle here: AppDelegate instance should exist for the duration of the life of the application.

**Pull request merged**